wled00/colors.cpp

```diff
@@ -91,6 +91,122 @@ void setRandomColor(byte* rgb)
+  //generate 4 saturation and brightness value numbers
+  //only one saturation is allowed to be below 200 creating mostly vibrant colors
+  //only one brightness value number is allowed below 200, creating mostly bright palettes  
+
+  for (int i = 0; i < 3; i++) { //generate three high values
+    palettecolors[i].saturation = random8(200,255);
+    palettecolors[i].value = random8(220,255);
+  }
+  //allow one to be lower
+  palettecolors[3].saturation = random8(80,255);
+  palettecolors[3].value = random8(80,255);
+
+
+  //shuffle the arrays 
+  for (int i = 3; i > 0; i--) {
+    std::swap(palettecolors[i].saturation, palettecolors[random8(i + 1)].saturation);
+    std::swap(palettecolors[i].value, palettecolors[random8(i + 1)].value);
+  }
+
+  //now generate three new hues based off of the hue of the chosen current color
+  uint8_t basehue = palettecolors[keepcolorposition].hue;
+  uint8_t harmonics[3]; //hues that are harmonic but still a little random
+  uint8_t type = random8(5); //choose a harmony type
+
+  switch (type) {
+    case 0: // analogous
+      harmonics[0] = basehue + random8(30, 50);
+      harmonics[1] = basehue + random8(10, 30);
+      harmonics[2] = basehue - random8(10, 30);
+      break;
+
+    case 1: // triadic
+      harmonics[0] = basehue + 113 + random8(15);
+      harmonics[1] = basehue + 233 + random8(15);
+      harmonics[2] = basehue -7 + random8(15);
+      break;
+
+    case 2: // split-complementary
+      harmonics[0] = basehue + 145 + random8(10);
+      harmonics[1] = basehue + 205 + random8(10);
+      harmonics[2] = basehue - 5 + random8(10);
+      break;
+    
+    case 3: // square
+      harmonics[0] = basehue + 85 + random8(10);
+      harmonics[1] = basehue + 175 + random8(10);
+      harmonics[2] = basehue + 265 + random8(10);
+     break;
+
+    case 4: // tetradic 
+      harmonics[0] = basehue + 80 + random8(20);
+      harmonics[1] = basehue + 170 + random8(20);
+      harmonics[2] = basehue + random8(30)-15;
+     break;
+  }
+
+
+  if (random8() < 128) //50:50 chance of shuffeling hues or keep the color order
+  {
+    //shuffle the hues:
+    for (int i = 2; i > 0; i--) {
+      std::swap(harmonics[i], harmonics[random8(i + 1)]);
+    }
+  }
+
+  //now set the hues
+  int j = 0;
+  for (int i = 0; i < 4; i++) {
+    if(i==keepcolorposition) continue; //skip the base color
+    palettecolors[i].hue = harmonics[j];
+    j++;
+   }
+
+  bool makepastelpalette = false;
+  if (random8() < 25) {//~10% chance of desaturated 'pastel' colors
+     makepastelpalette = true;
+  }
+
+  //apply saturation & gamma correction
+  CRGB RGBpalettecolors[4];
+  for (int i = 0; i < 4; i++) {
+    if(makepastelpalette && palettecolors[i].saturation > 180) { 
+      palettecolors[i].saturation -= 160; //desaturate all four colors
+    }    
+    RGBpalettecolors[i] = (CRGB)palettecolors[i]; //convert to RGB
+    RGBpalettecolors[i] = gamma32((uint32_t)RGBpalettecolors[i]);
```

Comment:
This one is a bit tricky, because the uint32_t operator of CRGB returns _RGBA_ (A=alpha channel) and not RGBW (W=white channel). This might be a problem for sk6812.  So better to cut off the 'A', for example uint32_t(RGBpalettecolors[i]) & 0x00FFFFFF`

Reply:
ok, I was struggling a bit as I was looking for a function that could do this in HSV, but then decided not to mess with it and just convert to RGB but was then wondering why it is 32bit not 24bit. thx for pointing this out. will fix it.